File: ta4j-core/src/main/java/org/ta4j/core/analysis/cost/FixedTransactionCostModel.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.analysis.cost;

import java.util.Objects;
import org.ta4j.core.Position;
import org.ta4j.core.Trade;
import org.ta4j.core.TradeFill;
import org.ta4j.core.num.Num;

/**
 * With this cost model, the trading costs for opening or closing a position are
 * accrued through a constant fee per trade (i.e. a fixed fee per transaction).
 */
public class FixedTransactionCostModel implements CostModel {

    /** The fixed fee per {@link Trade trade}. */
    private final double feePerTrade;

    /**
     * Constructor for a fixed fee trading cost model.
     *
     * <pre>
     * Cost of opened {@link Position position}: (fixedFeePerTrade * 1)
     * Cost of closed {@link Position position}: (fixedFeePerTrade * 2)
     * </pre>
     *
     * @param feePerTrade the fixed fee per {@link Trade trade}
     */
    public FixedTransactionCostModel(double feePerTrade) {
        this.feePerTrade = feePerTrade;
    }

    /**
     * @param position     the position
     * @param currentIndex the current bar index (irrelevant for
     *                     {@code FixedTransactionCostModel})
     * @return the transaction cost of the single {@code position}
     */
    @Override
    public Num calculate(Position position, int currentIndex) {
        Trade entry = position.getEntry();
        if (entry != null && entry.getFuturesContract() != null) {
            return sumExecutedFillCosts(position, currentIndex);
        }
        final var numFactory = position.getEntry().getPricePerAsset().getNumFactory();
        Num multiplier = numFactory.one();
        if (position.isClosed()) {
            multiplier = numFactory.numOf(2);
        }
        return numFactory.numOf(feePerTrade).multipliedBy(multiplier);
    }

    /**
     * <b>Note:</b> A partially executed native trade charges only the fills already
     * executed no later than {@code currentIndex}.
     *
     * <p>
     * The fixed fee is applied once per executed fill, because the venue charges
     * per contract execution rather than per entry/exit trade.
     * </p>
     */
    private Num sumExecutedFillCosts(Position position, int currentIndex) {
        Trade entry = position.getEntry();
        Num total = entry.getPricePerAsset().getNumFactory().zero();
        total = total.plus(sumFillCosts(entry, currentIndex));
        Trade exit = position.getExit();
        if (exit != null) {
            total = total.plus(sumFillCosts(exit, currentIndex));
        }
        return total;
    }

    private Num sumFillCosts(Trade trade, int currentIndex) {
        Num total = trade.getPricePerAsset().getNumFactory().zero();
        for (TradeFill fill : Trade.executionFillsOf(trade)) {
            if (fill.index() < 0 || fill.index() > currentIndex) {
                continue;
            }
            Num fillCost = feePerTrade == 0d || !fill.hasRecordedFees() ? calculate(fill) : fill.fee();
            total = total.plus(total.getNumFactory().numOf(fillCost.getDelegate()));
        }
        return total;
    }

    /**
     * @return the transaction cost of the single {@code position}
     */
    @Override
    public Num calculate(Position position) {
        Trade entry = position.getEntry();
        if (entry != null && entry.getFuturesContract() != null) {
            return sumExecutedFillCosts(position, Integer.MAX_VALUE);
        }
        return this.calculate(position, 0);
    }

    /**
     * <b>Note:</b> Both {@code price} and {@code amount} are irrelevant as the fee
     * in {@code FixedTransactionCostModel} is always the same.
     *
     * @return {@link #feePerTrade}
     */
    @Override
    public Num calculate(Num price, Num amount) {
        return price.getNumFactory().numOf(feePerTrade);
    }

    /**
     * Charges {@link #feePerTrade} once per native execution fill.
     *
     * @param fill the execution fill
     * @return the trading cost of {@code fill}
     * @since 0.25.1
     */
    @Override
    public Num calculate(TradeFill fill) {
        Objects.requireNonNull(fill, "fill");
        if (fill.futuresContract() == null) {
            return calculate(fill.price(), fill.amount());
        }
        return fill.price().getNumFactory().numOf(feePerTrade);
    }

    @Override
    public boolean equals(CostModel otherModel) {
        boolean equality = false;
        if (this.getClass().equals(otherModel.getClass())) {
            equality = ((FixedTransactionCostModel) otherModel).feePerTrade == this.feePerTrade;
        }
        return equality;
    }
}

File: ta4j-core/src/main/kotlin/org/ta4j/core/analysis/cost/FixedTransactionCostModel.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.analysis.cost

import org.ta4j.core.Position
import org.ta4j.core.Trade
import org.ta4j.core.num.*

open class FixedTransactionCostModel
/**
 * Constructor for a fixed fee trading cost model.
 *
 * Open [position][Position] cost: (fixedFeePerTrade * 1) Closed
 * [position][Position] cost: (fixedFeePerTrade * 2)
 *
 * @param feePerTrade the fixed fee per [trade][Trade])
 */(
    /**
     * Cost per [trade][Trade].
     */
    private val feePerTrade: Double
) : CostModel {
    /**
     * Calculates the transaction cost of a position.
     *
     * @param position     the position
     * @param finalIndex current bar index (irrelevant for the
     * FixedTransactionCostModel)
     * @return the absolute position cost
     */
    override fun calculate(position: Position, finalIndex: Int): Num? {
        val pricePerAsset = position.entry!!.pricePerAsset
        var multiplier = pricePerAsset!!.numOf(1)
        if (position.isClosed) {
            multiplier = pricePerAsset.numOf(2)
        }
        return pricePerAsset.numOf(feePerTrade).times(multiplier)
    }

    /**
     * Calculates the transaction cost of a position.
     *
     * @param position the position
     * @return the absolute position cost
     */
    override fun calculate(position: Position): Num? {
        return this.calculate(position, 0)
    }

    /**
     * Calculates the transaction cost based on the price and the amount (both
     * irrelevant for the FixedTransactionCostModel as the fee is always the same).
     *
     * @param price  the price per asset
     * @param amount number of traded assets
     */
    override fun calculate(price: Num?, amount: Num?): Num? {
        return price!!.numOf(feePerTrade)
    }

    /**
     * Evaluate if two models are equal
     *
     * @param other model to compare with
     */
    override fun equals(other: Any?): Boolean {
        var equality = false
        if (this.javaClass == other!!.javaClass) {
            equality = (other as FixedTransactionCostModel?)!!.feePerTrade == feePerTrade
        }
        return equality
    }

    companion object {
        private const val serialVersionUID = 3486293523619720786L
    }
}